modules/tectonic/resources/tectonic.sh

```diff
@@ -34,120 +24,87 @@ kubectl() {
 			return
 		fi
 
-		echo "kubectl failed, retrying in 5 seconds"
-		sleep 5
-	done
-}
-
-wait_for_crd() {
-	set +e
-	i=0
-
-	echo "Waiting for CRD $2"
-	until $KUBECTL -n "$1" get customresourcedefinition "$2"
-	do
-		i=$((i+1))
-		echo "CRD $2 not available yet, retrying in 5 seconds ($i)"
+		(>&2 echo "kubectl $* failed. Retrying in 5 seconds...")
 		sleep 5
 	done
-	set -e
 }
 
 wait_for_pods() {
-	set +e
-	echo "Waiting for pods in namespace $1"
+	echo "Waiting for pods in namespace $1..."
 	while true
 	do
-		out=$($KUBECTL -n "$1" get po -o custom-columns=STATUS:.status.phase,NAME:.metadata.name)
-		status=$?
+		out=$(kubectl --namespace "$1" get pods --output custom-columns=STATUS:.status.phase,NAME:.metadata.name --no-headers=true)
 		echo "$out"
 
-		if [ "$status" -ne "0" ]
+		# make sure kubectl returns at least one status
+		if [ "$(wc --lines <<< "$out")" -eq 0 ]
 		then
-			echo "kubectl command failed, retrying in 5 seconds"
+			echo "No pods were found. Waiting for 5 seconds..."
 			sleep 5
 			continue
 		fi
 
-		# make sure kubectl does not return "no resources found"
-		if [ "$(echo "$out" | tail -n +2 | grep -c '^')" -eq 0 ]
-		then
-			echo "no resources were found, retrying in 5 seconds"
-			sleep 5
-			continue
-		fi
-
-		stat=$(echo "$out"| tail -n +2 | grep -v '^Running')
-		if [ -z "$stat" ]
+		if ! grep --invert-match '^Running' <<< "$out"
```

Comment:
[Only `-v` is in POSIX][1].

And you might want to dump `grep`'s stdout into `/dev/null`, unless you're intentionally writing it to let the caller know where we're at.

[1]: http://pubs.opengroup.org/onlinepubs/9699919799/utilities/grep.html

Reply:
Printing the progress is an intentional side-effect.